idgen/idgen.go

```diff
@@ -0,0 +1,147 @@
+package idgen
+
+import (
+	"fmt"
+	"math/bits"
+	"sync"
+
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	kbucket "github.com/libp2p/go-libp2p-kbucket"
+)
+
+// HydraIdentityGenerator is a shared balanced ID generator.
+var HydraIdentityGenerator = NewBalancedIdentityGenerator()
+
+// BalancedIdentityGenerator is a facility for generating IPFS identities (i.e. IPFS private keys),
+// whose corresponding DHT keys are highly balanced, compared to just generating random keys.
+// Balancing is accomplished using "the power of two choices" paradigm:
+// https://www.eecs.harvard.edu/~michaelm/postscripts/mythesis.pdf
+//
+// New identities are generated by calling AddBalanced. BalancedIdentityGenerator remembers
+// generated identities, in order to ensure balance for future identities.
+// Generated identities can be removed using Remove.
+//
+// BalancedIdentityGenerator maintains the invariant that all identities, presently in its memory,
+// form an almost-perfectly balanced set.
+type BalancedIdentityGenerator struct {
+	sync.Mutex
+	xorTrie *XorTrie
+	count   int
+}
+
+// NewBalancedIdentityGenerator creates a new balanced identity generator.
+func NewBalancedIdentityGenerator() *BalancedIdentityGenerator {
+	return &BalancedIdentityGenerator{
+		xorTrie: NewXorTrie(),
+	}
+}
+
+// AddUnbalanced is used for testing purposes. It generates a purely random identity,
```

Comment:
I believe the intent is to use it for testing outside of unit tests as well

Reply:
Yeah, I would keep it here for now, because this XorTrie may end up being its own useful library for other purposes. At bare minimum it allows you to experiment/simulate with mixing balanced and unbalanced nodes (e.g. hydra and non-hydra nodes), so it may have a scope beyond testing.
Maybe I should fix the comment instead.